Neos.Flow/Classes/Reflection/ReflectionService.php

```diff
@@ -1447,12 +1447,27 @@ protected function reflectClassMethod($className, MethodReflection $method)
             $this->classesByMethodAnnotations[$annotationClassName][$className][] = $methodName;
         }
 
-        $returnType = $method->getDeclaredReturnType();
-        if ($returnType !== null && !in_array($returnType, ['self', 'null', 'callable', 'void', 'iterable', 'object', 'mixed']) && !TypeHandling::isSimpleType($returnType)) {
-            $returnType = '\\' . $returnType;
-        }
-        if ($method->isDeclaredReturnTypeNullable()) {
-            $returnType = '?' . $returnType;
+        $returnType= $method->getDeclaredReturnType();
+        $applyLeadingSlashIfNeeded = function (string $type) {
+            if (!in_array($type, ['self', 'null', 'callable', 'void', 'iterable', 'object', 'mixed'])
+                && !TypeHandling::isSimpleType($type)
+                && !TypeHandling::isLiteralType($type)
+            ) {
+                return '\\' . $type;
+            }
+            return $type;
+        };
+        if ($returnType !== null) {
+            if (TypeHandling::isUnionType($returnType)) {
```

Comment:
Basically this is okay as is for now.
I do think we should move all this type handling stuff back into PHP standard reflection API land though. So instead of our custom `getDeclaredReturnType()` that returns the stringified type, then parse it back if it was a union/intersection type, we should use the `getReturnType()` method as primary source of information directly
https://www.php.net/manual/en/class.reflectiontype.php
That can quickly become complex, because we might want to wrap that into a new method that falls back to docblock parsing info, which would mean we'd need to use own inherited `ReturnType` classes that we can manually instanciate. Or maybe there is a simpler intermediate step.

Reply:
Maybe we can parse the docblocks to ReflectionTypes ...

Reply:
Reading again parsing docblocks to reflections is exactly the core of your suggestion. So i agree to that.